### PR TITLE
Add min_acquire_period to DetectorDriverPart

### PR DESCRIPTION
**Pull request by GDYendell**
This allows defining of a minimum acheivable acquire period for a detector in the top level yaml config

Outstanding points:
- I think the way I have updated the `ADPilatus_runnable_block` to take `min_acquire_period` and pass it through to `Detector DriverPart` means we now have to define it in the top level yaml config - is there a way to make this optional or have a sensible default (does it support, e.g., `min_acquire_period: $(min_acquire_period=0.002)`)?

### CODE REVIEW

File: malcolm/modules/ADCore/parts/detectordriverpart.py [hunk 1]
```diff
@@ -234,6 +239,17 @@ def setup(self, registrar: PartRegistrar) -> None:
             "attributesToCapture", self.extra_attributes, self.set_extra_attributes
         )
 
+    @add_call_types
+    def on_validate(
+        self,
+        generator: scanning.hooks.AGenerator,
+    ) -> None:
+        if self.min_acquire_period is not None:
+            assert generator.duration >= self.min_acquire_period, (
```
[the remark this answers is not in the document]
Comment (BenBradnick) [reply malcolm/modules/ADCore/parts/detectordriverpart.py:248]: I made a mistake here - DetectorChildPart handles frames_per_step > 1 already - it creates a new generator with the duration divided down which is then passed to the DetectorDriverPart.

**Comment by GDYendell** [comment]
I have tested the latest version on I07 and it still works as before.=, but there is an issue.

With `duration = minimum_acquire_period = 0.05`:

```
malcolm.core.errors.BadValueError: Validate of BL07I-ML-PILAT-01 failed: AssertionError: Duration 0.049602 is less than minimum acquire period 0.05s
```

but if I use the same values in the unittests it works fine (`duration == minimum_acquire_period). Where is the precision being lost?

**Comment by BenBradnick** [comment]
The pmacChildPart will modify the generator duration based on aligning it with the servo frequency - you could verify with a print (either here or in RunnableController when tweaks are applied): https://github.com/dls-controls/pymalcolm/blob/6438903dbda4d12d9909d286d257dc94c98ec045/malcolm/modules/pmac/parts/pmacchildpart.py#L172

**Comment by GDYendell** [comment]
OK cool. It seems the duration will only ever be reduced, so we can just keep the check as is. It does seem to already be changed when it gets to that point (I added "Changing duration from ...")

```
In [1]: root: Generator passed to Compound init
root: {'typeid': 'scanpointgenerator:generator/StaticPointGenerator:1.0', 'size': 1000, 'axes': []}
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: BL07I-ML-SCAN-02: ReportStatusHook: Starting hook
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: ReportStatusHook: Child pulseTrigger returned Array([DetectorMutiframeInfo('BL07I-ML-PILAT-01')]) after 0.0007283687591552734s. Returning...
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: BL07I-ML-SCAN-02: ValidateHook: Starting hook
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: ValidateHook: Child simultaneousAxes returned Array([]) after 0.00395965576171875s. Still waiting for ['PANDA1', 'PANDA2', 'BRICK', 'PILATUS']
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-PILAT-01: BL07I-ML-PILAT-01: ReportStatusHook: Starting hook
malcolm.modules.ADPandABlocks.controllers.pandarunnablecontroller.PandARunnableController.BL07I-ML-PANDA-01: BL07I-ML-PANDA-01: ReportStatusHook: Starting hook
malcolm.modules.ADPandABlocks.controllers.pandarunnablecontroller.PandARunnableController.BL07I-ML-PANDA-02: BL07I-ML-PANDA-02: ReportStatusHook: Starting hook
malcolm.modules.pmac.parts.pmacchildpart.PmacChildPart.BRICK: Might change duation here
root: Generator passed to Compound init
root: {'typeid': 'scanpointgenerator:generator/StaticPointGenerator:1.0', 'size': 1000, 'axes': []}
malcolm.modules.pmac.parts.pmacchildpart.PmacChildPart.BRICK: Changing duration from 0.049602 to 0.049602
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: ValidateHook: Child BRICK returned Array([ParameterTweakInfo('generator', CompoundGenerator(generators=Array([StaticPointGenerator(size=1000, axes=Array([]))]), excluders=Array([]), mutators=Array([]), duration=0.049602, continuous=False, delay_after=0.0))]) after 0.012318611145019531s. Still waiting for ['PANDA1', 'PANDA2', 'PILATUS']
```

**Comment by GDYendell** [comment]
It is the pilatus DEADTIME reducing the exposure.

I think this is fine if exposure only gets reduced. It just might give a slightly unclear error message. If a user wants the exposure to be exactly 0.05 they need to add the deadtime to the duration.

**Comment by BenBradnick** [comment]
PR #336 should help out with that - I will probably ask you to test it at some point, as it should allow you to define an exposure time you want (e.g. 0.05) and pass a zero value for duration. It may still need some consideration for when the brick is giving you every trigger (I have only been using start of row triggering on K11).

**Comment by GDYendell** [comment]
> It is the pilatus DEADTIME reducing the exposure.

No it isn't. But the conclusion is the same:

```
malcolm.modules.pmac.parts.pmacchildpart.PmacChildPart.BRICK: Calculated duration 0.049602 does not match CompoundGenerator(generators=Array([StaticPointGenerator(size=1000, axes=Array([]))]), excluders=Array([]), mutators=Array([]), duration=0.05, continuous=False, delay_after=0.0)
root: Generator passed to Compound init
root: {'typeid': 'scanpointgenerator:generator/StaticPointGenerator:1.0', 'size': 1000, 'axes': []}
malcolm.modules.pmac.parts.pmacchildpart.PmacChildPart.BRICK: Changing CompoundGenerator(generators=Array([StaticPointGenerator(size=1000, axes=Array([]))]), excluders=Array([]), mutators=Array([]), duration=0.05, continuous=False, delay_after=0.0) -> CompoundGenerator(generators=Array([StaticPointGenerator(size=1000, axes=Array([]))]), excluders=Array([]), mutators=Array([]), duration=0.049602, continuous=False, delay_after=0.0)
malcolm.modules.scanning.controllers.runnablecontroller.RunnableController.BL07I-ML-SCAN-02: ValidateHook: Child BRICK returned Array([ParameterTweakInfo('generator', CompoundGenerator(generators=Array([StaticPointGenerator(size=1000, axes=Array([]))]), excluders=Array([]), mutators=Array([]), duration=0.049602, continuous=False, delay_after=0.0))]) after 0.015224218368530273s. Still waiting for ['PANDA2', 'PANDA1', 'PILATUS']
```

**Pull request merged**